### ISSUE TITLE
Restore cursor position if selection is canceled

### ISSUE THREAD

**Issue by liebkne** [zed-industries/zed#8588]
### Check for existing issues

- [X] Completed

### Describe the feature

Currently when "selecting larger/smaller syntax node", the cursor is moved to the end of the selection. If one then cancel the selection, the cursor is left at the end of the original selection.

I wonder if it is better to restore the cursor position before starting selection automatically, because I find that when I cancel selection, I need to move back to the original position quite often which is a bit inconvenient especially when the selection is expanded larger than the visible window. "GoBack" does not work in this situation, and I personally prefer it's done automatically when selection is canceled to avoid polluting the "GoBack" stack.

How do you think?

**Comment by SomeoneToIgnore** [comment]
Oh, sounds like a duplicate of your other issue? https://github.com/zed-industries/zed/issues/8105
That would be great to add and I think might not be very hard so start, at least.

* create a test

https://github.com/zed-industries/zed/blob/225dd0f9a0aff5c21f3f2208548438acfeaa9b5a/crates/editor/src/editor_tests.rs#L3712-L3739

show how to init a test buffer with text, then track carets in it `ˇ` and, below selection ranges: `def«ˇg»hi`
Something similar can be added to test the `SelectAll` -> `Cancel` test with carets, and there will be problems:

* we need to understand, whether we allow carets inside of selections: right now, at least tests, expect `ˇ` to be either next to the left (`«ˇ`) or right (`ˇ»`) edge selection: it will panic if we try to assert `«fsˇfs»` things

Maybe, it's fine to keep the selection mechanism working here this way, all existing selections work so that they change the caret position. Maybe it's not, some design/justification is needed there.

* finally, adjust https://github.com/zed-industries/zed/blob/f374f12df68384edd82f024f9dbef9fc53bacdbf/crates/editor/src/editor.rs#L2349-L2353 or, rather, the `try_cancel` method to actually place the caret back.

On that place, there's `self.selection_history` available that should have the previous caret, but `try_cancel`'s `self` has no access there — we need to think whether we should that previous caret to `MutableSelectionsCollection` or take it from editor's `SelectionHistory`, or do something else clever — so far it seems that picking the old one from the selection stack is fine, as long as we do not remove it from there.

**Comment by liebkne** [comment]
😢I just can not believe I forget I opened a similar issue for it…

**Comment by nikensss** [comment]
Hi! I would like to try to tackle this issue with @mrnugget next wednesday. He opened up a calendly request to have a 1h session with him during next week, and I managed to make a booking on Wednesday 13th at 15h. Would that be ok? I will be looking at this until then.

**Comment by nikensss** [comment]
Hi @failable, I have been able to reproduce the issue. Just to confirm, when you say:
>cancel the selection

Do you mean as in pressing the `<ESC>` key? And last question, are we talking about the selections when doing `Expand selection` (ie, `Control` + `Shift` + `Right`)?
![Screenshot 2024-03-12 at 13 47 36](https://github.com/zed-industries/zed/assets/40516571/a026099a-d4b0-4490-8fdf-d6902ea474b6)

**Comment by failable** [comment]
> Do you mean as in pressing the <ESC> key? 

Yes. Not sure if it is `menu::Cancel` or `editor::Cancel` single both they are bound to `esc`.

> And last question, are we talking about the selections when doing Expand selection (ie, Control + Shift + Right)?

Yes, I mean commands like `"editor::SelectLarger/SmallerSyntaxNode"` and `editor::SelectAll`, as mentioned in the old issue https://github.com/zed-industries/zed/issues/8105

**Comment by mrnugget** [comment]
So @nikensss and I just looked at this and we're not sure whether we should do this, or how exactly it should work if we were to do it. No other editor (we looked at Sublime, VS Code, Neovim) does restore the cursor position. I also think it might be a "breaking" change for a lot of people, or unexpected behavior.

My thinking here is that everywhere else the selection follows the cursor: if you start a selection and then move your cursor in a way that it gets added to the navigation history, then you can jump back after canceling the selection, it's independent of whether selection was on or not, the selection just follows the cursor.

The proposal here would change this though in that navigation-with-selection would now be a different thing from navigation-without-selection. And there's nothing wrong with that, I'm just not sure how to cleanly add this right now.

**Comment by SomeoneToIgnore** [comment]
To note, JetBrains IDEs restore the caret position after select all is cancelled, but they behave differently around the selection logic.

As the video tries to show, even when Select All is triggered, the caret is blinking on the previous place, so restoring there is simple.
Zed on the test level panics when the caret is not on the edge of the selection.

https://github.com/zed-industries/zed/assets/2690773/b122558d-0e38-4958-a999-584f988fb982
